File: android/app/src/main/java/com/shakilclark/lgremote/tv/Commands.kt
package com.shakilclark.lgremote.tv

import com.shakilclark.lgremote.connection.TvConnectionManager
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.emptyFlow
import kotlinx.coroutines.flow.map
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put

/** Volume/mute snapshot from the TV's getVolume subscription. */
data class VolumeState(val volume: Int? = null, val muted: Boolean? = null)

/** An external input/source (US7). */
data class TvInput(val id: String, val label: String, val connected: Boolean = false)

/** An installed app reported by the TV's launch points (dynamic app loader). */
data class TvApp(val id: String, val title: String, val iconUrl: String? = null)

/**
 * Control commands over SSAP (US2/US3/US6/US7; contracts/ssap-protocol.md). Stateless except the
 * play/pause toggle, which tracks the last media action (webOS has no reliable play-state query).
 */
class Commands(private val manager: TvConnectionManager) {

    private var playing: Boolean? = null

    suspend fun volumeUp() { manager.request("ssap://audio/volumeUp") }
    suspend fun volumeDown() { manager.request("ssap://audio/volumeDown") }

    suspend fun channelUp() { manager.request("ssap://tv/channelUp") }
    suspend fun channelDown() { manager.request("ssap://tv/channelDown") }

    suspend fun setMute(mute: Boolean) {
        manager.request("ssap://audio/setMute", buildJsonObject { put("mute", mute) })
    }

    suspend fun play() { manager.request("ssap://media.controls/play"); playing = true }
    suspend fun pause() { manager.request("ssap://media.controls/pause"); playing = false }

    /** Toggle; default to pause when the state is unknown (matches the 001 behaviour). */
    suspend fun playPause() { if (playing == true) pause() else play() }

    suspend fun rewind() { manager.request("ssap://media.controls/rewind") }
    suspend fun fastForward() { manager.request("ssap://media.controls/fastForward") }
    suspend fun stop() { manager.request("ssap://media.controls/stop"); playing = false }

    /**
     * Live foreground-app id (004) — the reliable "what's on" signal, present for every app/source
     * (e.g. `netflix`, `youtube.leanback.v4`, `com.webos.app.hdmi1`). Subscribable.
     */
    fun foregroundAppUpdates(): Flow<String?> =
        (manager.subscribe("ssap://com.webos.applicationManager/getForegroundAppInfo") ?: emptyFlow())
            .map { it["appId"]?.jsonPrimitive?.contentOrNull?.takeIf { id -> id.isNotBlank() } }

    /**
     * Live media play-state (004), where the app registers with the webOS media server — many apps
     * (Netflix, etc.) report an empty array, so this is best-effort; identity comes from
     * [foregroundAppUpdates].
     */
    fun mediaUpdates(): Flow<MediaForeground> =
        (manager.subscribe("ssap://com.webos.media/getForegroundAppInfo") ?: emptyFlow())
            .map { parseMediaForeground(it) }

    /** Live volume/mute stream; empty when not connected. */
    fun volumeUpdates(): Flow<VolumeState> =
        (manager.subscribe("ssap://audio/getVolume") ?: emptyFlow()).map { parseVolume(it) }

    // --- US6: app shortcuts (dynamic loader) ---

    /** All installed apps from the TV's launch points, in the TV's own order (dynamic app loader). */
    suspend fun listApps(): List<TvApp> =
        parseLaunchPoints(manager.request("ssap://com.webos.applicationManager/listLaunchPoints"))

    /** Launch an app by its resolved launch-point id. */
    suspend fun launchAppId(id: String): Boolean {
        val res = manager.request("ssap://system.launcher/launch", buildJsonObject { put("id", id) })
        return res["returnValue"]?.jsonPrimitive?.booleanOrNull ?: true
    }

    /**
     * Open the TV's own settings on-screen (bottom-bar ⚙). Launches the webOS settings app via the
     * same launcher used for app shortcuts — resolved by title where possible, falling back to the
     * well-known settings app id. Returns false if the TV won't launch it.
     */
    suspend fun openSettings(): Boolean {
        val id = runCatching {
            val res = manager.request("ssap://com.webos.applicationManager/listLaunchPoints")
            resolveLaunchPointIdByTitle(res, "Settings")
        }.getOrNull() ?: SETTINGS_APP_ID
        val res = manager.request("ssap://system.launcher/launch", buildJsonObject { put("id", id) })
        return res["returnValue"]?.jsonPrimitive?.booleanOrNull ?: true
    }

    private companion object {
        /** webOS settings app id; used if the TV's launch-point list has no "Settings" title. */
        const val SETTINGS_APP_ID = "com.palm.app.settings"
    }

    // --- US7: inputs ---

    suspend fun listInputs(): List<TvInput> =
        parseInputs(manager.request("ssap://tv/getExternalInputList"))

    suspend fun setInput(inputId: String) {
        manager.request("ssap://tv/switchInput", buildJsonObject { put("inputId", inputId) })
    }
}

/**
 * The TV returns volume either flat `{ volume, muted }` or nested
 * `{ volumeStatus: { volume, muteStatus } }` depending on firmware — handle both (observed on
 * the real TV in 001). Pure + unit-tested.
 */
fun parseVolume(payload: JsonObject): VolumeState {
    val flatVol = payload["volume"]?.jsonPrimitive?.intOrNull
    val flatMute = (payload["muted"] ?: payload["mute"])?.jsonPrimitive?.booleanOrNull
    val nested = payload["volumeStatus"] as? JsonObject
    val volume = flatVol ?: nested?.get("volume")?.jsonPrimitive?.intOrNull
    val muted = flatMute ?: nested?.get("muteStatus")?.jsonPrimitive?.booleanOrNull
    return VolumeState(volume, muted)
}

/**
 * Parse `listLaunchPoints` → the installed apps in the TV's order. Keeps id + title + the raw icon
 * string (may be a fetchable URL or a non-servable path — the UI loads it and falls back to a tile).
 * Pure + unit-tested.
 */
fun parseLaunchPoints(payload: JsonObject): List<TvApp> {
    val points = (payload["launchPoints"] as? kotlinx.serialization.json.JsonArray) ?: return emptyList()
    return points.mapNotNull { el ->
        val obj = el as? JsonObject ?: return@mapNotNull null
        val id = obj["id"]?.jsonPrimitive?.contentOrNull ?: return@mapNotNull null
        val title = obj["title"]?.jsonPrimitive?.contentOrNull ?: id
        val icon = obj["icon"]?.jsonPrimitive?.contentOrNull?.takeIf { it.isNotBlank() }
        TvApp(id = id, title = title, iconUrl = icon)
    }
}

/** Match a launch-point by title (case-insensitive) → its id. Null if not present. Pure. */
fun resolveLaunchPointIdByTitle(payload: JsonObject, title: String): String? {
    val points = (payload["launchPoints"] as? kotlinx.serialization.json.JsonArray) ?: return null
    return points.firstNotNullOfOrNull { el ->
        val obj = el as? JsonObject ?: return@firstNotNullOfOrNull null
        val pointTitle = obj["title"]?.jsonPrimitive?.contentOrNull
        val id = obj["id"]?.jsonPrimitive?.contentOrNull
        if (pointTitle?.equals(title, ignoreCase = true) == true) id else null
    }
}

/** Parse getExternalInputList `{ devices: [{ id, label, connected }] }`. Pure + unit-tested. */
fun parseInputs(payload: JsonObject): List<TvInput> {
    val devices = (payload["devices"] as? kotlinx.serialization.json.JsonArray) ?: return emptyList()
    return devices.mapNotNull { el ->
        val obj = el as? JsonObject ?: return@mapNotNull null
        val id = obj["id"]?.jsonPrimitive?.contentOrNull ?: return@mapNotNull null
        val label = obj["label"]?.jsonPrimitive?.contentOrNull ?: id
        val connected = obj["connected"]?.jsonPrimitive?.booleanOrNull ?: false
        TvInput(id = id, label = label, connected = connected)
    }
}

File: backend/src/tv/commands.ts
// Maps app commands → SSAP requests on the live TV socket, and keeps volume/mute in the
// ConnectionState via a getVolume subscription. Contract: contracts/tv-protocol.md.

import type { TVClient } from "./client.js";
import type { ConnectionStateMachine } from "../state/connection.js";
import type { AppKey, ControlCommand, ControlResult, TVInput } from "../types.js";

type Conn = NonNullable<TVClient["raw"]>;
type Pointer = { send(type: string, payload?: Record<string, unknown>): void; close(): void };

const POINTER_URI = "ssap://com.webos.service.networkinput/getPointerInputSocket";
const LIST_APPS_URI = "ssap://com.webos.applicationManager/listLaunchPoints";
const LAUNCH_URI = "ssap://system.launcher/launch";
const LIST_INPUTS_URI = "ssap://tv/getExternalInputList";
const SWITCH_INPUT_URI = "ssap://tv/switchInput";

// Fallback ids when listLaunchPoints can't be matched (varies by webOS version).
const WELL_KNOWN_APP_IDS: Record<AppKey, string> = {
  youtube: "youtube.leanback.v4",
  netflix: "netflix",
};

export class Commands {
  private lastPlaying = true; // best-effort; we don't get true media state for free
  private pointerPromise: Promise<Pointer> | null = null;
  private launchPoints: Array<{ id: string; title: string }> = [];

  constructor(
    private readonly client: TVClient,
    private readonly state: ConnectionStateMachine,
  ) {
    // (Re)subscribe to volume on every (re)connect so the UI mirrors the TV.
    this.client.onConnected = (conn) => {
      this.pointerPromise = null; // force re-acquire of the pointer socket after reconnect
      this.launchPoints = [];
      // Cache the TV's installed apps so we can resolve shortcut ids by title.
      conn.request(LIST_APPS_URI, null, (err, res) => {
        if (!err && Array.isArray(res?.launchPoints)) {
          this.launchPoints = res.launchPoints
            .filter((p: { id?: string; title?: string }) => p.id && p.title)
            .map((p: { id: string; title: string }) => ({ id: p.id, title: p.title }));
        }
      });
      conn.subscribe("ssap://audio/getVolume", (err, res) => {
        if (err || !res) return;
        // webOS varies: newer nests under volumeStatus{volume,muteStatus}; older is flat.
        const vs = (res.volumeStatus ?? res) as Record<string, unknown>;
        const patch: { volume?: number; muted?: boolean } = {};
        if (typeof vs.volume === "number") patch.volume = vs.volume;
        if (typeof vs.muteStatus === "boolean") patch.muted = vs.muteStatus;
        else if (typeof res.muted === "boolean") patch.muted = res.muted;
        if (Object.keys(patch).length) this.state.update(patch);
      });
    };
  }

  async send(cmd: ControlCommand): Promise<ControlResult> {
    const conn = this.client.raw;
    if (!this.client.isConnected || !conn) {
      return { result: "failed", message: "TV not connected", state: this.state.get() };
    }
    try {
      switch (cmd.type) {
        case "volumeUp":
          await request(conn, "ssap://audio/volumeUp");
          break;
        case "volumeDown":
          await request(conn, "ssap://audio/volumeDown");
          break;
        case "setMute": {
          const mute = Boolean(cmd.params?.mute);
          await request(conn, "ssap://audio/setMute", { mute });
          this.state.update({ muted: mute });
          break;
        }
        case "playPause": {
          const play = !this.lastPlaying;
          await request(conn, play ? "ssap://media.controls/play" : "ssap://media.controls/pause");
          this.lastPlaying = play;
          break;
        }
        case "nav": {
          const button = cmd.params?.button;
          if (!button) return { result: "failed", message: "nav requires a button", state: this.state.get() };
          const pointer = await this.getPointer(conn);
          pointer.send("button", { name: button });
          break;
        }
        case "launchApp": {
          const app = cmd.params?.app;
          if (!app) return { result: "failed", message: "launchApp requires an app", state: this.state.get() };
          const id = this.resolveAppId(app);
          const res = (await request(conn, LAUNCH_URI, { id })) as { returnValue?: boolean } | undefined;
          if (res && res.returnValue === false) {
            return { result: "failed", message: `${app} isn't available on this TV`, state: this.state.get() };
          }
          break;
        }
        case "setInput": {
          const inputId = cmd.params?.inputId;
          if (!inputId) return { result: "failed", message: "setInput requires an inputId", state: this.state.get() };
          const res = (await request(conn, SWITCH_INPUT_URI, { inputId })) as { returnValue?: boolean } | undefined;
          if (res && res.returnValue === false) {
            return { result: "failed", message: "Couldn't switch input", state: this.state.get() };
          }
          break;
        }
        default:
          return { result: "failed", message: "Unsupported command", state: this.state.get() };
      }
      return { result: "acknowledged", state: this.state.get() };
    } catch (e) {
      return { result: "failed", message: (e as Error).message, state: this.state.get() };
    }
  }

  /** List the TV's external inputs/sources (US7). Empty when not connected. */
  async listInputs(): Promise<TVInput[]> {
    const conn = this.client.raw;
    if (!this.client.isConnected || !conn) return [];
    const res = (await request(conn, LIST_INPUTS_URI)) as
      | { devices?: Array<{ id?: string; label?: string }> }
      | undefined;
    return (res?.devices ?? [])
      .filter((d): d is { id: string; label?: string } => typeof d.id === "string")
      .map((d) => ({ id: d.id, label: d.label ?? d.id }));
  }

  /** Resolve a shortcut app to its TV app id: match a launch-point title, else fall back. */
  private resolveAppId(app: AppKey): string {
    const match = this.launchPoints.find((p) => p.title.toLowerCase().includes(app));
    return match?.id ?? WELL_KNOWN_APP_IDS[app];
  }

  /** Acquire (and cache) the pointer-input socket used for D-pad/OK/Back/Home buttons. */
  private getPointer(conn: Conn): Promise<Pointer> {
    if (!this.pointerPromise) {
      this.pointerPromise = new Promise<Pointer>((resolve, reject) => {
        conn.getSocket(POINTER_URI, (err, sock) => {
          if (err || !sock) {
            this.pointerPromise = null;
            reject(err ?? new Error("Could not open pointer input socket"));
          } else {
            resolve(sock as Pointer);
          }
        });
      });
    }
    return this.pointerPromise;
  }
}

function request(conn: Conn, uri: string, payload?: Record<string, unknown>): Promise<unknown> {
  return new Promise((resolve, reject) => {
    conn.request(uri, payload ?? null, (err, res) => (err ? reject(err) : resolve(res)));
  });
}
